go-controller/pkg/clustermanager/userdefinednetwork/template/net-attach-def-template_test.go

```diff
@@ -2,12 +2,373 @@ package template
+
+	udnv1 "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1"
 )
 
 var _ = Describe("NetAttachDefTemplate", func() {
-	It("should fail", func() {
+	const udnTypeName = "UserDefinedNetwork"
+
+	var udnApiVersion = udnv1.SchemeGroupVersion.String()
+
+	DescribeTable("should fail given",
+		func(spec *udnv1.UserDefinedNetworkSpec) {
+			udn := &udnv1.UserDefinedNetwork{
+				ObjectMeta: metav1.ObjectMeta{Namespace: "mynamespace", Name: "test-net", UID: "1"},
+				Spec:       *spec,
+			}
+			_, err := RenderNetAttachDefManifest(udn)
+			Expect(err).To(HaveOccurred())
+		},
+		Entry("invalid topology: topology layer2 & layer3 config",
+			&udnv1.UserDefinedNetworkSpec{Topology: udnv1.NetworkTopologyLayer2, Layer3: &udnv1.Layer3Config{}},
+		),
+		Entry("invalid topology: topology layer2 & localnet config",
+			&udnv1.UserDefinedNetworkSpec{Topology: udnv1.NetworkTopologyLayer2, LocalNet: &udnv1.LocalNetConfig{}},
+		),
+		Entry("invalid topology: topology layer3 & layer2 config",
+			&udnv1.UserDefinedNetworkSpec{Topology: udnv1.NetworkTopologyLayer3, Layer2: &udnv1.Layer2Config{}},
+		),
+		Entry("invalid topology: topology layer3 & localnet config",
+			&udnv1.UserDefinedNetworkSpec{Topology: udnv1.NetworkTopologyLayer3, LocalNet: &udnv1.LocalNetConfig{}},
+		),
+		Entry("invalid topology: topology localnet & layer2 config",
+			&udnv1.UserDefinedNetworkSpec{Topology: udnv1.NetworkTopologyLocalNet, Layer2: &udnv1.Layer2Config{}},
+		),
+		Entry("invalid topology: topology localnet & layer3 config",
+			&udnv1.UserDefinedNetworkSpec{Topology: udnv1.NetworkTopologyLocalNet, Layer3: &udnv1.Layer3Config{}},
+		),
+		Entry("topology=localnet & role=primary",
+			&udnv1.UserDefinedNetworkSpec{
+				Topology: udnv1.NetworkTopologyLocalNet,
+				LocalNet: &udnv1.LocalNetConfig{
+					Role: udnv1.NetworkRolePrimary,
+				},
+			},
+		),
+		Entry("invalid layer2 subnets",
+			&udnv1.UserDefinedNetworkSpec{
+				Topology: udnv1.NetworkTopologyLayer2,
+				Layer2: &udnv1.Layer2Config{
+					Subnets: udnv1.DualStackCIDRs{"abc"},
+				},
+			},
+		),
+		Entry("invalid localnet subnets",
+			&udnv1.UserDefinedNetworkSpec{
+				Topology: udnv1.NetworkTopologyLayer2,
+				LocalNet: &udnv1.LocalNetConfig{
+					Subnets: udnv1.DualStackCIDRs{"abc"},
+				},
+			},
+		),
+		Entry("invalid layer3 cluster-subnet",
+			&udnv1.UserDefinedNetworkSpec{
+				Topology: udnv1.NetworkTopologyLayer3,
+				Layer3: &udnv1.Layer3Config{
+					Subnets: []udnv1.Layer3Subnet{{CIDR: "!", HostSubnet: 16}},
+				},
+			},
+		),
+		Entry("invalid layer3 host-subnet mask",
+			&udnv1.UserDefinedNetworkSpec{
+				Topology: udnv1.NetworkTopologyLayer3,
+				Layer3: &udnv1.Layer3Config{
+					Subnets: []udnv1.Layer3Subnet{
+						{CIDR: "10.10.0.0/24", HostSubnet: -1},
+					},
+				},
+			},
+		),
+		Entry("layer3 host-subnet mask is smaller then cluster-subnet mask",
```

Comment:
Its validated down the stack calling `NewNetInfo` [[1] ](https://github.com/ovn-org/ovn-kubernetes/blob/e02ff8cbedfb6d7bf490d23d1b38fe6c5e62c565/go-controller/pkg/util/multi_network.go#L663) by `parseSubnets` here [[2]  ](https://github.com/ovn-org/ovn-kubernetes/blob/e02ff8cbedfb6d7bf490d23d1b38fe6c5e62c565/go-controller/pkg/config/utils.go#L99)

I added this test to have indication in case `NewNetworkInfo` changes, so such change "wont go silently".

[1] https://github.com/ovn-org/ovn-kubernetes/blob/e02ff8cbedfb6d7bf490d23d1b38fe6c5e62c565/go-controller/pkg/util/multi_network.go#L663
[2] https://github.com/ovn-org/ovn-kubernetes/blob/e02ff8cbedfb6d7bf490d23d1b38fe6c5e62c565/go-controller/pkg/config/utils.go#L99